File: desktop/src/features/control-room/lib/agentHealth.ts
export type HealthStatus = "green" | "amber" | "red";
export type AssuranceStatus = "complete" | "partial" | "insufficient";
export type DimensionState = "pass" | "fail" | "unknown";

export type HealthDimension = {
  state: DimensionState;
  evidence: string[];
};

export type AgentHealthRecord = {
  id: string;
  name: string;
  operationalStatus: HealthStatus;
  assuranceStatus: AssuranceStatus;
  dimensions: Record<string, HealthDimension>;
};

export type HealthRecord = {
  id: string;
  name: string;
  status: HealthStatus;
  detail: string;
};

export type AgentHealthSnapshot = {
  schemaVersion: "mac-agent-health/v1";
  generatedAt: string;
  operationalStatus: HealthStatus;
  assuranceStatus: AssuranceStatus;
  assuranceGaps: string[];
  source: {
    status: "fresh" | "stale" | "invalid";
    estate?: { observedAt: string; ageSeconds: number };
    agents?: { observedAt: string; ageSeconds: number };
  };
  nodes: HealthRecord[];
  agents: AgentHealthRecord[];
  components: HealthRecord[];
  issues: string[];
};

const STATUS = new Set<HealthStatus>(["green", "amber", "red"]);
const ASSURANCE = new Set<AssuranceStatus>([
  "complete",
  "partial",
  "insufficient",
]);
const DIMENSION_STATE = new Set<DimensionState>(["pass", "fail", "unknown"]);

function object(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${label} must be an object`);
  }
  return value as Record<string, unknown>;
}

function string(value: unknown, label: string): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new Error(`${label} must be a non-empty string`);
  }
  return value;
}

function strings(value: unknown, label: string): string[] {
  if (
    !Array.isArray(value) ||
    !value.every((item) => typeof item === "string")
  ) {
    throw new Error(`${label} must be a string array`);
  }
  return value;
}

function status(value: unknown): HealthStatus {
  if (!STATUS.has(value as HealthStatus)) {
    throw new Error("Unsupported health status");
  }
  return value as HealthStatus;
}

function assurance(value: unknown): AssuranceStatus {
  if (!ASSURANCE.has(value as AssuranceStatus)) {
    throw new Error("Unsupported assurance status");
  }
  return value as AssuranceStatus;
}

function records(value: unknown, label: string): HealthRecord[] {
  if (!Array.isArray(value)) throw new Error(`${label} must be an array`);
  return value.map((item) => {
    const raw = object(item, label);
    return {
      id: string(raw.id, `${label}.id`),
      name: string(raw.name, `${label}.name`),
      status: status(raw.status),
      detail: string(raw.detail, `${label}.detail`),
    };
  });
}

function agents(value: unknown): AgentHealthRecord[] {
  if (!Array.isArray(value)) throw new Error("agents must be an array");
  return value.map((item) => {
    const raw = object(item, "agent");
    const rawDimensions = object(raw.dimensions, "agent.dimensions");
    const dimensions = Object.fromEntries(
      Object.entries(rawDimensions).map(([name, value]) => {
        const dimension = object(value, `agent.dimensions.${name}`);
        if (!DIMENSION_STATE.has(dimension.state as DimensionState)) {
          throw new Error("Unsupported health dimension state");
        }
        return [
          name,
          {
            state: dimension.state as DimensionState,
            evidence: strings(
              dimension.evidence,
              `agent.dimensions.${name}.evidence`,
            ),
          },
        ];
      }),
    );
    return {
      id: string(raw.id, "agent.id"),
      name: string(raw.name, "agent.name"),
      operationalStatus: status(raw.operationalStatus),
      assuranceStatus: assurance(raw.assuranceStatus),
      dimensions,
    };
  });
}

export function parseAgentHealthSnapshot(value: unknown): AgentHealthSnapshot {
  const raw = object(value, "agent health");
  if (raw.schemaVersion !== "mac-agent-health/v1") {
    throw new Error("Unsupported MAC agent-health snapshot");
  }
  const source = object(raw.source, "source");
  const sourceStatus = string(source.status, "source.status");
  if (!["fresh", "stale", "invalid"].includes(sourceStatus)) {
    throw new Error("Unsupported source status");
  }
  const parseObservation = (name: "estate" | "agents") => {
    if (!source[name]) return undefined;
    const evidence = object(source[name], `source.${name}`);
    return {
      observedAt: string(evidence.observedAt, `source.${name}.observedAt`),
      ageSeconds:
        typeof evidence.ageSeconds === "number" ? evidence.ageSeconds : 0,
    };
  };
  return {
    schemaVersion: "mac-agent-health/v1",
    generatedAt: string(raw.generatedAt, "generatedAt"),
    operationalStatus: status(raw.operationalStatus),
    assuranceStatus: assurance(raw.assuranceStatus),
    assuranceGaps: strings(raw.assuranceGaps, "assuranceGaps"),
    source: {
      status: sourceStatus as "fresh" | "stale" | "invalid",
      estate: parseObservation("estate"),
      agents: parseObservation("agents"),
    },
    nodes: records(raw.nodes, "nodes"),
    agents: agents(raw.agents),
    components: records(raw.components, "components"),
    issues: strings(raw.issues, "issues"),
  };
}

export function agentHealthEndpoint(relayUrl: string): URL {
  const endpoint = new URL(relayUrl);
  if (endpoint.protocol === "wss:") endpoint.protocol = "https:";
  else if (endpoint.protocol === "ws:") endpoint.protocol = "http:";
  else if (!["https:", "http:"].includes(endpoint.protocol)) {
    throw new Error("Unsupported relay protocol");
  }
  endpoint.pathname = "/api/mac-agent-health/v1";
  endpoint.search = "";
  endpoint.hash = "";
  return endpoint;
}

export async function loadAgentHealth({
  relayUrl,
  signal,
}: {
  relayUrl: string;
  signal?: AbortSignal;
}): Promise<AgentHealthSnapshot> {
  const response = await fetch(agentHealthEndpoint(relayUrl), {
    cache: "no-store",
    headers: { Accept: "application/json" },
    signal,
  });
  if (!response.ok) {
    throw new Error(`Agent health is unavailable (${response.status})`);
  }
  return parseAgentHealthSnapshot(await response.json());
}

File: mobile/lib/features/control_room/agent_health.dart
enum HealthStatus { green, amber, red }

enum AssuranceStatus { complete, partial, insufficient }

enum HealthDimensionState { pass, fail, unknown }

class HealthDimension {
  final HealthDimensionState state;
  final List<String> evidence;

  const HealthDimension({required this.state, required this.evidence});

  factory HealthDimension.fromJson(Map<String, dynamic> json) {
    return HealthDimension(
      state: switch (json['state']) {
        'pass' => HealthDimensionState.pass,
        'fail' => HealthDimensionState.fail,
        'unknown' => HealthDimensionState.unknown,
        _ => throw const FormatException('Unsupported health dimension state'),
      },
      evidence: _strings(json['evidence']),
    );
  }
}

class AgentHealthRecord {
  final String id;
  final String name;
  final HealthStatus operationalStatus;
  final AssuranceStatus assuranceStatus;
  final Map<String, HealthDimension> dimensions;

  const AgentHealthRecord({
    required this.id,
    required this.name,
    required this.operationalStatus,
    required this.assuranceStatus,
    required this.dimensions,
  });

  factory AgentHealthRecord.fromJson(Map<String, dynamic> json) {
    final rawDimensions = _map(json['dimensions']);
    return AgentHealthRecord(
      id: _text(json['id']),
      name: _text(json['name']),
      operationalStatus: _status(json['operationalStatus']),
      assuranceStatus: _assurance(json['assuranceStatus']),
      dimensions: rawDimensions.map(
        (key, value) => MapEntry(key, HealthDimension.fromJson(_map(value))),
      ),
    );
  }
}

class HealthRecord {
  final String id;
  final String name;
  final HealthStatus status;
  final String detail;

  const HealthRecord({
    required this.id,
    required this.name,
    required this.status,
    required this.detail,
  });

  factory HealthRecord.fromJson(Map<String, dynamic> json) {
    return HealthRecord(
      id: _text(json['id']),
      name: _text(json['name']),
      status: _status(json['status']),
      detail: _text(json['detail']),
    );
  }
}

class AgentHealthSnapshot {
  final DateTime? generatedAt;
  final HealthStatus operationalStatus;
  final AssuranceStatus assuranceStatus;
  final List<String> assuranceGaps;
  final DateTime? estateObservedAt;
  final List<HealthRecord> nodes;
  final List<AgentHealthRecord> agents;
  final List<HealthRecord> components;
  final List<String> issues;

  const AgentHealthSnapshot({
    required this.generatedAt,
    required this.operationalStatus,
    required this.assuranceStatus,
    required this.assuranceGaps,
    required this.estateObservedAt,
    required this.nodes,
    required this.agents,
    required this.components,
    required this.issues,
  });

  factory AgentHealthSnapshot.fromJson(Map<String, dynamic> json) {
    if (json['schemaVersion'] != 'mac-agent-health/v1') {
      throw const FormatException('Unsupported MAC agent-health snapshot');
    }
    final source = _map(json['source']);
    return AgentHealthSnapshot(
      generatedAt: DateTime.tryParse(_text(json['generatedAt'])),
      operationalStatus: _status(json['operationalStatus']),
      assuranceStatus: _assurance(json['assuranceStatus']),
      assuranceGaps: _strings(json['assuranceGaps']),
      estateObservedAt: DateTime.tryParse(
        _text(_map(source['estate'])['observedAt']),
      ),
      nodes: _maps(json['nodes']).map(HealthRecord.fromJson).toList(),
      agents: _maps(json['agents']).map(AgentHealthRecord.fromJson).toList(),
      components: _maps(json['components']).map(HealthRecord.fromJson).toList(),
      issues: _strings(json['issues']),
    );
  }
}

Uri agentHealthUri(String relayUrl) {
  final relay = Uri.parse(relayUrl);
  final scheme = switch (relay.scheme) {
    'wss' => 'https',
    'ws' => 'http',
    'https' => 'https',
    'http' => 'http',
    _ => throw const FormatException('Invalid community relay URL'),
  };
  return relay.replace(
    scheme: scheme,
    path: '/api/mac-agent-health/v1',
    query: null,
    fragment: null,
  );
}

HealthStatus _status(dynamic value) => switch (value) {
  'green' => HealthStatus.green,
  'amber' => HealthStatus.amber,
  'red' => HealthStatus.red,
  _ => throw const FormatException('Unsupported health status'),
};

AssuranceStatus _assurance(dynamic value) => switch (value) {
  'complete' => AssuranceStatus.complete,
  'partial' => AssuranceStatus.partial,
  'insufficient' => AssuranceStatus.insufficient,
  _ => throw const FormatException('Unsupported assurance status'),
};

Map<String, dynamic> _map(dynamic value) {
  return value is Map ? Map<String, dynamic>.from(value) : {};
}

List<Map<String, dynamic>> _maps(dynamic value) {
  return value is List ? value.map(_map).toList() : [];
}

List<String> _strings(dynamic value) {
  if (value is! List || value.any((item) => item is! String)) {
    throw const FormatException('Expected a string array');
  }
  return value.cast<String>();
}

String _text(dynamic value) => value is String ? value : '';
